src/storage/mutate/AddVerticesProcessor.cpp

```diff
@@ -30,78 +30,66 @@ void AddVerticesProcessor::process(const cpp2::AddVerticesRequest& req) {
     }
 
     CHECK_NOTNULL(kvstore_);
-    if (indexes_.empty()) {
-        std::for_each(partVertices.begin(), partVertices.end(), [&](auto& pv) {
-            auto partId = pv.first;
-            const auto& vertices = pv.second;
-            std::vector<kvstore::KV> data;
-            std::for_each(vertices.begin(), vertices.end(), [&](auto& v) {
-                const auto& tags = v.get_tags();
-                std::for_each(tags.begin(), tags.end(), [&](auto& tag) {
-                    VLOG(3) << "PartitionID: " << partId << ", VertexID: " << v.get_id()
-                            << ", TagID: " << tag.get_tag_id() << ", TagVersion: " << version;
-                    auto key = NebulaKeyUtils::vertexKey(partId, v.get_id(),
-                                                         tag.get_tag_id(), version);
-                    data.emplace_back(std::move(key), std::move(tag.get_props()));
-                    if (FLAGS_enable_vertex_cache && vertexCache_ != nullptr) {
-                        vertexCache_->evict(std::make_pair(v.get_id(), tag.get_tag_id()));
-                        VLOG(3) << "Evict cache for vId " << v.get_id()
-                                << ", tagId " << tag.get_tag_id();
-                    }
-                });
+    std::for_each(partVertices.begin(), partVertices.end(), [&](auto& pv) {
+        std::vector<kvstore::KV> data;
+        data.reserve(128);
+        std::vector<std::tuple<VertexID, TagID, std::string>> cacheData;
+        if (FLAGS_enable_vertex_cache && vertexCache_ != nullptr) {
+            cacheData.reserve(128);
+        }
+        auto partId = pv.first;
+        const auto& vertices = pv.second;
+        std::for_each(vertices.begin(), vertices.end(), [&](auto& v) {
+            const auto& tags = v.get_tags();
+            std::for_each(tags.begin(), tags.end(), [&](auto& tag) {
+                VLOG(3) << "PartitionID: " << partId << ", VertexID: " << v.get_id()
+                        << ", TagID: " << tag.get_tag_id() << ", TagVersion: " << version;
+                auto key = NebulaKeyUtils::vertexKey(partId, v.get_id(),
+                                                     tag.get_tag_id(), version);
+                if (FLAGS_enable_vertex_cache && vertexCache_ != nullptr) {
+                    cacheData.emplace_back(v.get_id(), tag.get_tag_id(), tag.get_props());
+                }
+                data.emplace_back(std::move(key), std::move(tag.get_props()));
```

Comment:
One question,Does it cause duplicate indexes to be inserted? for example : 
insert vertex ({vid1_tag1_version1, prop1}, {vid1_tag1_version2, prop2}...)

Reply:
When you insert {vid1_tag1_version, prop2}, the index generated by ({vid1_tag1_version, prop1} will be deleted as our logic.

Reply:
Good one.  I miss the logic.  Maybe we could combine the inserts before calling kvstore::asyncPut